### PR TITLE
Capture console logs from user codes

### PR DESCRIPTION
**Pull request by lohanidamodar**
All runtimes except Swift are working
## Changes
-  Update the model of response from runtimes
- Skipped console logs capturing test in Swift
- Deno only has 1.21 and 1.24 removed other versions
- Removed node 15.5 and 17.0 runtimes
- Removed dart 2.12, 2.13 and 2.14

### CODE REVIEW

File: runtimes/node-17.0/server.js [hunk 1]
```diff
@@ -38,8 +43,10 @@ const server = micro(async (req, res) => {
             await userFunction(request, response);
         }
     } catch (e) {
-        send(res, 500, e.code === 'MODULE_NOT_FOUND' ? "Code file not found." : e.stack || e);
+        send(res, 500, {stdout: logs, stderr: e.code === 'MODULE_NOT_FOUND' ? "Code file not found." : e.stack || e});
     }
+    logs.length = 0;
+    console.log = console.stdlog;
 });
```
Comment (Meldiron) [inline runtimes/node-17.0/server.js:50]: This sounds unsafe. Why not `logs = []`, and why are we changing value of `console.log` but not `console.error` and others?

File: runtimes/node-17.0/server.js [hunk 2]
```diff
@@ -17,9 +17,14 @@ const server = micro(async (req, res) => {
         payload: body.payload ?? '',
     };
 
+    console.stdlog = console.log.bind(console);
+    logs = [];
+    console.log = function(){
+        logs.push(Array.from(arguments));
```
Comment (Meldiron) [inline runtimes/node-17.0/server.js:23]: Soo if I low 2 things, are they part of one array, or are they separated?

`console.log(1,3)` is `[1,3]` or [[1,3]]`? Also not sure which one is better.

File: runtimes/node-17.0/server.js [hunk 3]
```diff
@@ -17,9 +17,14 @@ const server = micro(async (req, res) => {
         payload: body.payload ?? '',
     };
 
+    console.stdlog = console.log.bind(console);
```
Comment (Meldiron) [inline runtimes/node-17.0/server.js:20]: can this be some our variable? Sounds scarry to set variable on `console` object

File: tests/node-17.0/tests.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ module.exports = async (req, res) => {
     const payload = JSON.parse(req.payload === '' ? '{}' : req.payload);
 
     const todo = await fetch(`https://jsonplaceholder.typicode.com/todos/${payload.id ?? 1}`).then(r => r.json());
-
+    console.log('stdout', 'other outputs');
```
Comment (Meldiron) [inline tests/node-17.0/tests.js:19]: In test, can we add multiple logs? also multiple types, maybe even try to log object. For all, lets have test check in PHP

File: runtimes/node-17.0/server.js (hunk 2 above)
Comment (lohanidamodar) [reply runtimes/node-17.0/server.js:23]: if I don't do `Array.from(arguments)` `console.log(1,3)` becomes `{0: 1, 1: 3}`

File: tests/node-17.0/tests.js (hunk 4 above)
Comment (lohanidamodar) [reply tests/node-17.0/tests.js:19]: haven't gone to the test yet, still trying to get the runtimes right before moving on to test

File: runtimes/node-17.0/server.js (hunk 1 above)
Comment (lohanidamodar) [reply runtimes/node-17.0/server.js:50]: what should I capture, let me know, I just used the StackOverflow reference you provided me and implemented this.

File: runtimes/node-17.0/server.js (hunk 3 above)
Comment (lohanidamodar) [reply runtimes/node-17.0/server.js:20]: it can be, will update to that

File: runtimes/node-17.0/server.js (hunk 1 above)
Comment (Meldiron) [reply runtimes/node-17.0/server.js:50]: CleanShot 2022-04-27 at 09 19 05

There are a few of them: `debug()`, `error()`, `info()`, `log()`, `warn()`. We just need to capture all of them, the same way we do it for `log()`.

File: runtimes/node-17.0/server.js (hunk 3 above)
Comment (Meldiron) [reply runtimes/node-17.0/server.js:20]: Let's keep it for internal use

**Comment by Meldiron** [comment]
- If you log array/object, it does not show actual value
- Let's join with ` ` arguments, and then logs with `\n`

**Comment by lohanidamodar** [comment]
@Meldiron need another set of reviews

File: test.http [hunk 5]
```diff
@@ -0,0 +1,7 @@
+POST http://0.0.0.0:3000
+content-type: application/json
+X-internal-Challenge: test-secret-key
+
+{
+    "garbage": "garbage"
```
Comment (Meldiron) [inline test.http:6]: Maaaybe a different word here 😅

File: tests/dart-2.16/lib/tests.dart [hunk 6]
```diff
@@ -15,17 +15,17 @@ import 'package:dio/dio.dart' hide Response;
...
 Future<void> start(final req, final res) async {
-
   final payload = jsonDecode(req.payload == '' ? '{}' : req.payload);
 
   final id = payload['id'] ?? '1';
-  final todo = await Dio().get('https://jsonplaceholder.typicode.com/todos/$id');
-
+  final todo =
+      await Dio().get('https://jsonplaceholder.typicode.com/todos/$id');
+  print('stdout');
```
Comment (Meldiron) [inline tests/dart-2.16/lib/tests.dart:23]: Can we add more prints to test? primitive types + object + array. If behavior is different from other runtimes, it might be fine, we will do a specific test for that runtime. The main idea is to fully understand what will be returned.

File: test.http (hunk 5 above)
Comment (lohanidamodar) [reply test.http:6]: this is something to be deleted, i made this file for local testing. I took this from test files

**Comment by lohanidamodar** [comment]
@abnegate Swift still seems failing please check!

File: runtimes/dart-2.12/server.dart [hunk 7]
```diff
@@ -25,12 +32,34 @@ void main() async {
       );
 
       final response = Response();
-      await user_code.start(request, response);
-      return shelf.Response.ok(response.body);
+      await runZonedGuarded(
+        () async {
+          await user_code.start(request, response);
+        },
+        (e, stackTrace) => print('Oh noes! $e $stackTrace'),
```
Comment (Meldiron) [inline runtimes/dart-2.12/server.dart:39]: Oh noes on purpose?

Comment (Meldiron) [reply runtimes/dart-2.12/server.dart:39]: It's in like 5 places.

Comment (lohanidamodar) [reply runtimes/dart-2.12/server.dart:39]: copy paste 😄

Comment (lohanidamodar) [reply runtimes/dart-2.12/server.dart]: fixed

**Review by Meldiron — APPROVED** [review]
Looks great

File: .travis.yml [hunk 8]
```diff
@@ -53,24 +59,24 @@ env:
...
+    #   ENTRYPOINT=tests.ts
+    #   SERVER_PROCESS=server.ts
+    #   IMAGE=openruntimes/deno:${VERSION}-1.13
+    #   ARCH=linux/amd64
+    # - RUNTIME=deno-1.12
+    #   PHP_CLASS=Deno112
+    #   ENTRYPOINT=tests.ts
+    #   SERVER_PROCESS=server.ts
+    #   IMAGE=openruntimes/deno:${VERSION}-1.12
+    #   ARCH=linux/amd64
```
Comment (christyjacob4) [inline .travis.yml:79]: I believe this needs to be uncommented ?

File: runtimes/deno-1.21/server.ts [hunk 9]
```diff
@@ -21,14 +21,33 @@ app.use(async (ctx) => {
...
+        }
+    });
+    logs.push(args.join(" "));
+  }
+
   const response = {
     send: (text: string, status = 200) => {
       ctx.response.status = status;
-      ctx.response.body = text;
+      ctx.response.body = {response: text, stdout: logs.join('\n')};
```
Comment (christyjacob4) [inline runtimes/deno-1.21/server.ts:46]: In the other runtimes ( c++, dart ) , we are sending the response body as a json encoded string. Here we're directly sending a JSON. I don't think this is expected.

File: runtimes/dotnet-3.1/src/RuntimeResponse.cs [hunk 10]
```diff
@@ -28,7 +27,7 @@ namespace DotNetRuntime
             Dictionary<string, object?> data,
             int statusCode = 200)
         {
-            Data = JsonSerializer.Serialize(data);
+            Data = data;
```
Comment (christyjacob4) [inline runtimes/dotnet-3.1/src/RuntimeResponse.cs:30]: I'm seeing different patterns being followed for different runtimes.
In some cases we're sending serialised strings, and in some we're sending the JSON object as it is. 

Please test all the runtimes if this is working

File: runtimes/deno-1.21/server.ts (hunk 9 above)
Comment (christyjacob4) [reply runtimes/deno-1.21/server.ts:46]: Please test all the runtimes to **verify** this is working

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt [hunk 11]
```diff
@@ -28,9 +30,34 @@ suspend fun execute(ctx: Context) {
...
+    System.setErr(userErr)
+
     try {
-        ctx.result(codeWrapper.main(request, response).data)
+        val userResponse = codeWrapper.main(request, response)
+        val output = mutableMapOf(
+            "response" to userResponse.data,
+            "stdout" to outStream.toString()
+        )
+        ctx.json(output)
```
Comment (christyjacob4) [inline runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:49]: Doesn't the output need to be serialised?  We are doing it in Java, and the others

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt [hunk 12]
```diff
@@ -28,9 +30,34 @@ suspend fun execute(ctx: Context) {
...
     } catch (e: Exception) {
-        ctx.status(500).result(e.stackTraceToString())
+        e.printStackTrace()
+        val output = mutableMapOf(
+            "stdout" to outStream.toString(),
+            "stderr" to errStream.toString()
+        )
+        ctx.status(500).json(output)
```
Comment (christyjacob4) [inline runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:56]: similarly here as well

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt (hunk 11 above)
Comment (christyjacob4) [reply runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:49]: @abnegate  helped resolve the comment 👍

File: runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt (hunk 12 above)
Comment (christyjacob4) [reply runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt:56]: @abnegate  helped resolve the comment 👍

File: runtimes/swift-5.5/Sources/App/routes.swift [hunk 13]
```diff
@@ -110,19 +100,51 @@ func routes(_ app: Application) throws {
...
         }
 
+        setvbuf(stdout, nil, _IONBF, 0)
+        setvbuf(stderr, nil, _IONBF, 0)
+
+        let outPipe = Pipe()
+        let errPipe = Pipe()
+        var outString = ""
+        var errString = ""
+
```
Comment (christyjacob4) [inline runtimes/swift-5.5/Sources/App/routes.swift:110]: Since console logs are not working in swift, can we revert these changes in swift to simply return an empty string for `stdout`?

**Review by christyjacob4 — COMMENTED** [review]
PR looks great. Please validate the comments and uncomment the tests for deno and we're ready to go 👍

File: runtimes/deno-1.21/server.ts (hunk 9 above)
Comment (lohanidamodar) [reply runtimes/deno-1.21/server.ts:46]: I think we should add new runtimes in deno and discard the old ones. I'll look at the releases, the old ones I don't think it's worth spending much time on them, so I've commented those out.

File: runtimes/swift-5.5/Sources/App/routes.swift (hunk 13 above)
Comment (lohanidamodar) [reply runtimes/swift-5.5/Sources/App/routes.swift:110]: Well I'm okay either way. as it's not working

File: .travis.yml (hunk 8 above)
Comment (lohanidamodar) [reply .travis.yml:85]: I've rather added new Deno 1.24 runtime, don't think maintaining and updating all these old deno runtimes is fruitful. Deno has lot newer releases `deno 1.21` and `deno 1.24`. let me know what you think

File: runtimes/deno-1.21/server.ts (hunk 9 above)
Comment (christyjacob4) [reply runtimes/deno-1.21/server.ts:46]: In that case, let's remove the older ones and finalise this PR

File: .travis.yml (hunk 8 above)
Comment (christyjacob4) [reply .travis.yml:85]: yes, this is a tedious process maintaining a lot of similar releases and is not worth it... 
Let's remove the old ones and keep just the latest 1 and an LTS

File: runtimes/swift-5.5/Sources/App/routes.swift (hunk 13 above)
Comment (christyjacob4) [reply runtimes/swift-5.5/Sources/App/routes.swift:110]: @abnegate please address this when you get the time.

File: tests/Deno124.php [hunk 14]
```diff
@@ -6,6 +6,6 @@
 // PHP class: Deno114
```
Comment (lohanidamodar) [inline tests/Deno124.php:6]:
```suggestion
// PHP class: Deno124
```

**Review by christyjacob4 — APPROVED** [review]

**Comment by christyjacob4** [comment]
Please create a corresponding PR on appwrite/runtimes to fetch the new versions
